File: q2872/q2872_dfs.java
class Solution {
    private int maxComponentCount = 0;

    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        List<Integer>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int[] edge : edges) {
            int src = edge[0];
            int dst = edge[1];
            graph[src].add(dst);
            graph[dst].add(src);
        }

        dfsHelper(0, -1, graph, values, k);
        
        return maxComponentCount;
    }

    // dfsHelper returns the sum of the subtree with a given node
    private int dfsHelper(int currNode, int parentNode, List<Integer>[] graph, int[] nodeVals, int k) {
        int currSum = nodeVals[currNode] % k; 
        for (int neighbor : graph[currNode]) {
            if (neighbor != parentNode) {
                currSum += dfsHelper(neighbor, currNode, graph, nodeVals, k);
                currSum %= k;
            }
        }
        // Whenever a currSum is zero, we have found a subtree with combined sum divisible by k
        // Greedily remove the subtree into one of the valid components, and increment the count by 1
        if (currSum == 0) maxComponentCount++;
        return currSum;
    }
}

File: q2872/q2872_dfs.py
class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        self.graph = defaultdict(list)
        self.max_components = 0

        for u, v in edges:
            self.graph[u].append(v)
            self.graph[v].append(u)

        # Start DFS from node 0, with parent = -1
        self.dfs_helper(0, -1, values, k)
        return self.max_components
    
    def dfs_helper(self, curr, parent, values, k):
        # Compute subtree sum with root node at curr 
        curr_sum = values[curr]

        for nxt in self.graph[curr]:
            if nxt != parent:
                curr_sum += self.dfs_helper(nxt, curr, values, k)

        # If a parent tree sum % k = 0, and it contains a subtree sum % k = 0
        # Then both subtree component, and (bigger parent tree - subtree) component should be counted
        # Hence whenever curr_sum % k == 0, eagerly add 1 to max_components
        if curr_sum % k == 0:
            self.max_components += 1
            # Optionally terminating this recursion, return 0
            return 0
        
        return curr_sum
